File: app/src/main/java/com/niedzwiecki/przemyslguide/ui/placeDetails/PlaceDetailsActivity.java
package com.niedzwiecki.przemyslguide.ui.placeDetails;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.v4.view.ViewPager;
import android.support.v7.widget.AppCompatImageView;
import android.view.View;
import android.view.WindowManager;
import android.view.animation.AlphaAnimation;
import android.view.animation.Animation;
import android.view.animation.AnimationSet;
import android.widget.TextView;

import com.niedzwiecki.przemyslguide.R;
import com.niedzwiecki.przemyslguide.data.model.PlaceOfInterest;
import com.niedzwiecki.przemyslguide.ui.PlaceDetailViewPager;
import com.niedzwiecki.przemyslguide.ui.base.BaseActivity;
import com.niedzwiecki.przemyslguide.ui.main.MainActivity;
import com.niedzwiecki.przemyslguide.ui.maps.MapsActivity;
import com.niedzwiecki.przemyslguide.util.Utils;
import com.squareup.picasso.Picasso;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

/**
 * Created by niedzwiedz on 10.07.17.
 */

public class PlaceDetailsActivity extends BaseActivity {

    private static final String INTEREST_PLACE_KEY = "interestPlaceKey";

    @BindView(R.id.nameOfRibot)
    TextView nameTextView;

    @BindView(R.id.description)
    TextView descriptionTextView;

    @BindView(R.id.coverImage)
    AppCompatImageView coverImage;

    @BindView(R.id.mailTextView)
    TextView mailTextView;

    @BindView(R.id.viewPager)
    ViewPager viewPager;

    private PlaceOfInterest place;
    private PlaceDetailViewPager adapter;

    public static Intent getStartIntent(Context context, PlaceOfInterest place) {
        Intent intent = new Intent(context, PlaceDetailsActivity.class);
        intent.putExtra(MainActivity.INTEREST_PLACE_KEY, place);
        return intent;
    }

    @Override
    public void beforeViews() {
        super.beforeViews();
        activityComponent().inject(this);
    }

    @Override
    public void afterViews() {
        super.afterViews();
        ButterKnife.bind(this);
        setScreenFlags();
        fetchData();
        setData();
    }

    private void fetchData() {
        if (getIntent().hasExtra(MainActivity.INTEREST_PLACE_KEY)) {
            place = getIntent().getExtras().getParcelable(MainActivity.INTEREST_PLACE_KEY);
        }
    }

    @Override
    protected void afterViews(Bundle savedInstanceState) {
        super.afterViews(savedInstanceState);
        ButterKnife.bind(this);
        setScreenFlags();
        fetchData();
        setData();
        if (savedInstanceState != null) {
            restoreData(savedInstanceState);
        }
    }

    private void restoreData(Bundle savedInstanceState) {
        place = savedInstanceState.getParcelable(INTEREST_PLACE_KEY);
    }

    private void setScreenFlags() {
        getWindow().setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN,
                WindowManager.LayoutParams.FLAG_FULLSCREEN);
    }

    private void setData() {
        if (Utils.isEmpty(place.name)) {
            nameTextView.setVisibility(View.GONE);
        } else {
            nameTextView.setText(place.name);
        }

        if (Utils.isEmpty(place.description)) {
            descriptionTextView.setVisibility(View.GONE);
        } else {
            descriptionTextView.setText(place.description);
        }

        mailTextView.setVisibility(View.GONE);

        if (!Utils.isEmpty(place.image)) {
            Animation fadeIn = new AlphaAnimation(0, 1);
            fadeIn.setDuration(3000);
            AnimationSet animation = new AnimationSet(true);
            animation.addAnimation(fadeIn);
            coverImage.setAnimation(animation);
            if (Utils.isEmpty(place.image)) {
                return;
            }

            Picasso.with(this)
                    .load(place.image)
                    .resize(700, 700)
                    .centerCrop()
                    .into(coverImage);
        } else {
            coverImage.setVisibility(View.GONE);
        }

        adapter = new PlaceDetailViewPager();
        viewPager.setAdapter(adapter);
    }

    @Override
    public int contentId() {
        return R.layout.activity_place_details;
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putParcelable(INTEREST_PLACE_KEY, place);
    }

    @OnClick(R.id.fabButton)
    public void onFabButtonClick() {
        Intent intent = new Intent(MapsActivity.getStartIntent(getBaseContext(), place));
        startActivity(intent);
    }
}

File: app/src/main/java/com/niedzwiecki/przemyslguide/ui/placeDetails/PlaceDetailsActivity.kt
package com.niedzwiecki.przemyslguide.ui.placeDetails

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.WindowManager
import com.niedzwiecki.przemyslguide.R
import com.niedzwiecki.przemyslguide.data.model.PlaceOfInterest
import com.niedzwiecki.przemyslguide.databinding.ActivityPlaceDetailsBinding
import com.niedzwiecki.przemyslguide.ui.base.BaseActivity
import com.niedzwiecki.przemyslguide.ui.base.Navigator
import com.niedzwiecki.przemyslguide.ui.base.ViewModel
import com.niedzwiecki.przemyslguide.ui.gallery.GalleryPagerAdapter
import com.niedzwiecki.przemyslguide.ui.main.MainActivity
import com.niedzwiecki.przemyslguide.ui.maps.MapsActivity
import com.niedzwiecki.przemyslguide.util.Utils.convertDpToPixel
import com.squareup.picasso.Picasso

/**
 * Created by niedzwiedz on 10.07.17.
 */

class PlaceDetailsActivity : BaseActivity() {

    private var place: PlaceOfInterest? = null
    private lateinit var placeAdapter: GalleryPagerAdapter

    override fun beforeViews() {
        super.beforeViews()
        setDataBindingEnabled(true)
    }

    override fun afterViews() {
        super.afterViews()
        viewDataBinding.viewModel = getViewModel()
        setScreenFlags()
        fetchData()
        setMapAndOnClick()
        toolbar.setTitle(" ")
        setDataToViewModel()
        placeAdapter = GalleryPagerAdapter(this)
        placeAdapter.setItems(place?.images)
        viewDataBinding.viewPager.adapter = placeAdapter
    }

    private fun setMapAndOnClick() {
        Picasso.with(this)
                .load("https://maps.googleapis.com/maps/api/staticmap?center=" + place?.lat +
                        "," + place?.lon +
                        "&zoom=14&size=800x450&maptype=roadmap&markers=color:blue%7Clabel:S%7C" + place?.lat +
                        "," + place?.lon + "&key=AIzaSyDthXCUL8-qPlt8F21aQRuMDGSt8WLveB8")
                .resize(convertDpToPixel(400f, this), convertDpToPixel(250f, this))
                .centerCrop()
                .into(viewDataBinding.staticMap)

        viewDataBinding.staticMap.setOnClickListener({
            if (place != null) {
                startMapActivity(place!!)
            }
        })
    }

    private fun fetchData() {
        if (intent.hasExtra(MainActivity.INTEREST_PLACE_KEY)) {
            place = intent.extras!!.getParcelable(MainActivity.INTEREST_PLACE_KEY)
        }
    }

    override fun afterViews(savedInstanceState: Bundle?) {
        super.afterViews(savedInstanceState)
        viewDataBinding.viewModel = getViewModel()
        setScreenFlags()
        fetchData()
        setDataToViewModel()
//        placeAdapter.setItems(place?.images)
        if (savedInstanceState != null) {
            restoreData(savedInstanceState)
        }
    }

    private fun restoreData(savedInstanceState: Bundle) {
        place = savedInstanceState.getParcelable(INTEREST_PLACE_KEY)
    }

    private fun setScreenFlags() {
        window.setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN,
                WindowManager.LayoutParams.FLAG_FULLSCREEN)
    }

    private fun setDataToViewModel() {
        getViewModel().setData(place)
    }

    override fun contentId(): Int {
        return R.layout.activity_place_details
    }

    override fun getViewDataBinding(): ActivityPlaceDetailsBinding {
        return (super.getViewDataBinding() as? ActivityPlaceDetailsBinding)!!
    }

    override fun createViewModel(): ViewModel {
        return PlaceDetailsViewModel(dataManager)
    }

    override fun getViewModel(): PlaceDetailsViewModel {
        return super.getViewModel() as PlaceDetailsViewModel
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putParcelable(INTEREST_PLACE_KEY, place)
    }

    override fun moveForward(options: Navigator.Options?, vararg data: Any?) {
        super.moveForward(options, *data)
        when (options) {
            Navigator.Options.START_ACTIVITY_WITH_INTENT ->
                startMapActivity(data[0] as PlaceOfInterest)
        }
    }

    private fun startMapActivity(place: PlaceOfInterest) {
        val intent = Intent(MapsActivity.getStartIntent(baseContext, place))
        startActivity(intent)
    }

    companion object {

        private val INTEREST_PLACE_KEY = "interestPlaceKey"

        fun getStartIntent(context: Context, place: PlaceOfInterest): Intent {
            val intent = Intent(context, PlaceDetailsActivity::class.java)
            intent.putExtra(MainActivity.INTEREST_PLACE_KEY, place)
            return intent
        }
    }
}
